npep/npep-122.md

```diff
@@ -0,0 +1,194 @@
+
+From the User Stories:
+
+    Story 4: Create and Isolate multiple tenants in a cluster
+    
+    As a cluster admin, I want to build tenants in my cluster that are isolated from each other by default. 
+    Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant 
+    may own more than 1 Namespace.
+
+Elsewhere:
+
+    AdminNetworkPolicy Pass rules allows an admin to delegate security posture for certain traffic to the Namespace 
+    owners by overriding any lower precedence Allow or Deny rules. For example, intra-tenant traffic management can be 
+    delegated to tenant admins explicitly with the use of Pass rules.
+    
+So really, the only solidly-agreed-upon use case is that you should be able to create enforced isolation 
+between particular sets of namespaces.
+
+### Clarifying user stories
+
+#### Story 4.1: Create and Isolate multiple tenants in a cluster by default
+
+Here is the existing tenancy related user story from our website:
+
+    As a cluster admin, I want to build tenants in my cluster that are isolated from each other by default. 
+    Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant 
+    may own more than 1 Namespace.
+
+The wording used here has lead to much confusion. Specifically, "By default" part actually means this policy should 
+be at the BANP priority, and solves the "Engineering org vs Marketing org" use case where you want to keep orgs 
+from accidentally interfering with each other, but still want to allow them to communicate if needed. 
+
+In reality, the CR attached to this story defines strict tenancy, which should be a separate use case (see next section).
+
+To make the use case more obvious we can add some details to it like the following
+
+    As a cluster admin, I want to build tenants in my cluster that are isolated from each other by default.
+    This should prevent tenants from interfering with each other by accident, but may be overridden by namespace
+    owners when needed.
+    Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant 
+    may own more than 1 Namespace.
+
+#### Story 4.2: Create and Isolate multiple tenants in a cluster, strict isolation
+
+Strict tenancy is the "Coke vs Pepsi" sort of thing where you want each tenant to feel like it has its own cluster, 
+and be totally independent of the other tenants. We can write it down like this
+
+    As a cluster admin, I want to build tenants in my cluster that are isolated from each other and this can't
+    be overridden. This policy should make every tenant completely independent and isolated from another tenants.
+    Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant 
+    may own more than 1 Namespace.
+
+#### Story 4.3: Allow internal connections for tenants
```

Comment:
Hrm this one seems a bit duplicative to the "create and isolate" stories to me...

More specifically... 

> At the same time I want to setup an overridable deny-all policy to protect namespaces by default.
    This policy should make sure internal connectivity for a tenant is always allowed, in case there are
    lower-priority deny rules.

In this case (thinking in terms on ANP) you'd use an ANP to setup the tenants... and that ANP would implicity be a "higher-order" policy meaning even with an "overridable deny-all policy" traffic within tenants would always be allowed unless explicitly denied with another ANP or "passed" to namespace owners

Reply:
It is very similar indeed, but the difference is in the action: "Isolate" user story is focused on the Deny tenancy rules, and "Allow internal connections for tenants" is more about  Allow tenancy rules. Maybe they are too similar, lmk if you think I should remove this one

Reply:
> "Isolate" user story is focused on the Deny tenancy rules, and "Allow internal connections for tenants" is more about Allow tenancy rules.

IMO this is somewhat implicit in:

```
As a cluster admin, I want to build tenants in my cluster that are isolated from each other where this isolation
    can't be overridden by namespace owners.
```

i.e we're building hard tenants where the definition is tenants can't talk to each other but members of a tenant are of one group, and can communicate.

If you want to augment 4.2 to explicitly mention something like that I think that'd be alright, but I think Dan and I agree here so I'd remove 4.3

```
   As a cluster admin, I want to build tenants in my cluster that are isolated from each other where this isolation
    can't be overridden by namespace owners. This policy should make every tenant completely independent and isolated 
    from another tenants, while explicitly allowing same tenant members to communicate.
 ```

Reply:
let me try to explain this once again, because I don't think we are talking about the same use case

4.2 story just denies inter-tenant communication, but doesn't say anything about intra-tenant communication, e.g. you could write "deny from the other tenants" ANP, and at the same time define a NP in namespace A saying "deny from namespace B", where namespaces A and B are in the same tenant. That would limit intra-tenant communication, and ANP doesn't say anything about that, so traffic from namespace B will be dropped.

4.3 use case is aimed to prevent namespaces in the same tenant from denying each others connections (e.g. by applying any network policy and by mistake not including your tenant namespaces to the allow list). In that case even if same namespace A applies a "deny from ALL" network policy, namespace B will still be allowed by ANP.

I just want to make sure I've explained the use cases well enough, and then we can surely remove any extra parts or whole user stories.

Reply:
> 4.3 use case is aimed to prevent namespaces in the same tenant from denying each others connections

Is that a real use case? Why would an administrator want to enforce that for every tenant at the cluster level?

4.1 is "Engineering vs Marketing" and 4.2 is "Coke vs Pepsi". What is the explanation for this user story?

(Also, you should probably actually put "Engineering vs Marketing" and "Coke vs Pepsi" into the actual user stories. The user story should stand alone to explain the use case. We shouldn't have the user story _and_ also have a separate "here's what this really means". And in this case (4.3), I don't think there _is_ any "here's what this really means".)